[Kuang Long] iP

angkl0 opened the pull request:
# Duke 

>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke is a task manager that improves your productivity. It is:
   - Text-based
   - Easy to learn
   - Intuitive

All you need to do is:
   1. Download it from [here]
   2. Manage your tasks using the many features available
   3. And you do not need to worry about your tasks anymore! 🤯
 
More importantly, Duke is **FREE**!!!

Features available:
   - [X] Tasks are categorized into todo, deadline and event
   - [X] Add, delete and search for tasks
   - [X] Mark tasks as done once you have completed them

Here's the main method:
```
public static void main(String[] args) throws IOException {
        Duke duke = new Duke();
        duke.run();
}
```

Commit:
Add Gradle support

angkl0 committed:
level 1

angkl0 committed:
level-2

angkl0 committed:
level-3

angkl0 committed:
level-4

angkl0 committed:
level-5

angkl0 committed:
level-6

angkl0 committed:
Add capability to save tasks

angkl0 committed:
Modify storage of dates

angkl0 committed:
Merge branch 'branch-Level-7'
Add capability to save tasks

angkl0 committed:
Resolve merge conflict

angkl0 committed:
Add Storage,Ui,TaskList,Parser,Command and its associated subclasses

angkl0 committed:
Organise some files into packages

angkl0 committed:
Merge branch 'add-gradle-support'
Merge add-gradle-support

angkl0 committed:
Add some tests

angkl0 committed:
Add Javadoc

angkl0 committed:
Modify code to comply with coding standard

angkl0 committed:
Add find

angkl0 committed:
Merge branch 'Branch-A-CodingStandard'
Merge branch-A-CodingStandard

angkl0 committed:
Merge branch 'branch-A-JavaDoc'
Merge branch-A-JavaDoc

angkl0 committed:
Resolve merge conflict

angkl0 committed:
Modfiy main

Eclipse-Dominator commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+
+    }
+
+    public void run() throws IOException {
```
Perhaps consider adding javadoc for public methods

Eclipse-Dominator commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+import java.io.File;
+import java.io.IOException;
+import java.time.format.DateTimeParseException;
+import duke.commands.*;
```
> Imported classes should always be listed explicitly.

Perhaps consider importing the individual Commands

angkl0 replied:
Thanks for the detailed feedback given. I have implemented most of the suggestions you have given.

Eclipse-Dominator commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+        Boolean description_needed = command.equals("todo") || command.equals("deadline") || command.equals("event") ||
+                command.equals("mark") || command.equals("unmark") || command.equals("delete") || command.equals("find");
+        if (splitInput.length <= 1 && description_needed) {
+            throw new DukeException("Description of command is required.");
```
> In general, try to avoid any magic literals.

Perhaps considering making a MissingDescriptionException class that extends DukeException to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+        case "find":
+            return new FindCommand(detail);
+        default:
+            throw new DukeException(command + " is an invalid command.");
```
> In general, try to avoid any magic literals.

Perhaps considering making a InvalidCommandException class that extends DukeException to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,47 @@
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException, IOException {
+        String[] splitDetailDeadline = this.input.split("/by", 2);
+        if (splitDetailDeadline.length == 1) {
+            throw new DukeException("Description of deadline must be followed by /by then followed by time of deadline.");
```
> In general, try to avoid any magic literals.

Perhaps considering making a InvalidTaskFormatException class that extends DukeException to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+        if (this.input.matches("\\d+")) {
+            int deleteIndex = Integer.parseInt(this.input) - 1;
+            if (deleteIndex < 0 || deleteIndex >= taskList.length()) {
+                throw new DukeException(String.format("There is no task with index %d", deleteIndex + 1));
```
> In general, try to avoid any magic literals.

Perhaps considering making a InvalidIndexException class that extends DukeException with a parameter to take in the index value to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+                ui.print(deleteMessage, taskList);
+            }
+        } else {
+            throw new DukeException(this.input + " is not an integer.");
```
> In general, try to avoid any magic literals.

Perhaps considering making a InvalidIntegerException class that extends DukeException to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,47 @@
+    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException, IOException {
+        String[] splitDetailEvent = this.input.split("/at", 2);
+        if (splitDetailEvent.length == 1) {
+            throw new DukeException("Description of event must be followed by /at then followed by time of event.");
```
> In general, try to avoid any magic literals.

Perhaps considering making a InvalidTaskFormatException class that extends DukeException to avoid magic String?
Since there are multiple uses of it, perhaps you either have InvalidEventFormatException or you can allow the InvalidTaskFormatException to take parameters for the type task as well as the seperator.

Eclipse-Dominator commented on src/main/java/duke/commands/MarkCommand.java:
```diff
@@ -0,0 +1,48 @@
+            } else {
+                Task markTask = taskList.index(markIndex);
+                if (markTask.getIsDone()) {
+                    throw new DukeException("This task is already marked as done:\n" + markTask.toString());
```
> In general, try to avoid any magic literals.

Perhaps considering making a InvalidIndexException class that extends DukeException with a parameter to take in the index value to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,53 @@
+            }
+        }
+        if (hit.size() == 0) {
+            throw new DukeException("There are no matching results.");
```
> In general, try to avoid any magic literals.

Perhaps considering making a NoTaskMatchException class that extends DukeException with a parameter to take in the index value to avoid magic String?

Eclipse-Dominator commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+        String[] splitInput = input.split(" ", 2);
+        String command = splitInput[0].trim();
+        String detail = "";
+        Boolean description_needed = command.equals("todo") || command.equals("deadline") || command.equals("event") ||
```
> Boolean variables/methods should be named to sound like booleans
> Variable names must be in camelCase.

Perhaps consider using `hasDescription` as a variable name

Eclipse-Dominator replied:
> Avoid Complicated Expressions

Instead of checking the command, perhaps you check the negation since you already have a default case below. Since you are only checking for list perhaps use `boolean hasDescription = !command.equals("list");`
However to account for the checking of missing description, perhaps you can actually abstract out CommandWithDescription from the various commands and throw the missing description error there as an exception?

Eclipse-Dominator commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,48 @@
+package duke;
+
+import java.util.ArrayList;
+import duke.tasks.*;
```
>  Imported classes should always be listed explicitly.

Would you agree that you should change to import the individual tasks instead of using a wild card import

Eclipse-Dominator commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+     */
+    public static Task parseFromFile(String line) {
+        String[] line_split = line.split("]", 3);
+        String type = line_split[0];
+        String status = line_split[1];
+        String rest = line_split[2];
```
Uncaught `ArrayIndexOutOfBoundException`. 
While you did specify a length of 3, however, if the line is say `Test ] test`, it will not actually give you an array of length 3!
Perhaps you can add an assertion or check to make sure that the array is of length 3 before proceeding.
You might argue it is an save file, but there are always chances for the file to become corrupt or modified so I think this is something you should take note.

Eclipse-Dominator commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+     * @return Task parsed from input string.
+     */
+    public static Task parseFromFile(String line) {
+        String[] line_split = line.split("]", 3);
```
> Variable names must be in camelCase.

Perhaps would you consider changing to to `lineSplit` to fit the java coding standard?

Eclipse-Dominator commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,93 @@
+        } else {
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("hh:mm a 'on' dd/MM/yyyy");
+            if (type.equals("[D")) {
+                String[] rest_split = rest.split("by:", 2);
```
>Variable names must be in camelCase.

Would you agree to consider changing it to `restSplit` to fit the java coding standard?

Eclipse-Dominator commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,55 @@
+        if (length > 0) {
+            fw.write(taskList.index(0).toString() + "\n");
+            fw.close();
+            FileWriter second_fw = new FileWriter(this.filePath, true);
```
> Variable names must be in camelCase.

Would you consider changing this to `secondFw` to fit the java coding standard

Eclipse-Dominator reviewed (commented):
Good work implementing the javadocs and the logic. Perhaps you can consider abstracting out some methods/classes and it would definitely improve the readability. Also take note to follow the given java coding standard!

angkl0 committed:
Add GUI

angkl0 committed:
Add assertions

Added assertions to document important assertions that should hold at
various points of the code.

angkl0 committed:
Improve code quality

Based on feedback from peer review, I have improved code quality in the
aspects of avoiding magic strings and correcting variable names that do
not conform to the coding standard.

angkl0 committed:
Merge pull request #1 from angkl0/branch-A-Assertions

Add assertions

angkl0 committed:
Merge branch 'master' into branch-A-CodeQuality
Merge master into branch-A-CodeQuality

angkl0 committed:
Merge pull request #2 from angkl0/branch-A-CodeQuality

Improve code quality

angkl0 committed:
Add ability to recognize and deal with duplicate tasks

angkl0 committed:
Update README.md

angkl0 committed:
Fix compatibility issue with Windows and improve code quality

angkl0 committed:
Update README.md

angkl0 committed:
Improve code quality

angkl0 committed:
Add Javadoc for more methods
